miden-crypto/src/hash/algebraic_sponge/mod.rs

```diff
@@ -0,0 +1,244 @@
+        // don't need to apply any extra padding because the first capacity element contains a
+        // flag indicating the number of field elements constituting the last block when the latter
+        // is not divisible by `RATE_WIDTH`.
+        if rate_pos != 0 {
+            state[RATE_RANGE.start + rate_pos..RATE_RANGE.end].fill(ZERO);
+            Self::apply_permutation(&mut state);
+        }
+
+        // return the first 4 elements of the rate as hash result.
+        Word::new(state[DIGEST_RANGE].try_into().unwrap())
+    }
+
+    /// Returns a hash of two digests. This method is intended for use in construction of
+    /// Merkle trees and verification of Merkle paths.
+    fn merge(values: &[Word; 2]) -> Word {
+        // initialize the state by copying the digest elements into the rate portion of the state
+        // (8 total elements), and set the capacity elements to 0.
+        let mut state = [ZERO; STATE_WIDTH];
+        let it = Word::words_as_elements_iter(values.iter());
+        for (i, v) in it.enumerate() {
+            state[RATE_RANGE.start + i] = *v;
+        }
+
+        // apply the permutation and return the digest portion of the state
+        Self::apply_permutation(&mut state);
+        Word::new(state[DIGEST_RANGE].try_into().unwrap())
+    }
+
+    /// Returns a hash of many digests.
+    fn merge_many(values: &[Word]) -> Word {
+        let elements = Word::words_as_elements(values);
+        Self::hash_elements(elements)
+    }
+
+    /// Returns hash(`seed` || `value`). This method is intended for use in PRNG and PoW contexts.
+    fn merge_with_int(seed: Word, value: u64) -> Word {
+        // initialize the state as follows:
+        // - seed is copied into the first 4 elements of the rate portion of the state.
+        // - if the value fits into a single field element, copy it into the fifth rate element and
+        //   set the first capacity element to 5.
+        // - if the value doesn't fit into a single field element, split it into two field elements,
+        //   copy them into rate elements 5 and 6 and set the first capacity element to 6.
+        let mut state = [ZERO; STATE_WIDTH];
+        state[INPUT1_RANGE].copy_from_slice(seed.as_elements());
+        state[INPUT2_RANGE.start] = Felt::new(value);
+        if value < Felt::MODULUS {
+            state[CAPACITY_RANGE.start] = Felt::from(5_u8);
+        } else {
+            state[INPUT2_RANGE.start + 1] = Felt::new(value / Felt::MODULUS);
+            state[CAPACITY_RANGE.start] = Felt::from(6_u8);
+        }
+
+        // apply the permutation and return the digest portion of the rate
+        Self::apply_permutation(&mut state);
+        Word::new(state[DIGEST_RANGE].try_into().unwrap())
+    }
+
+    // DOMAIN IDENTIFIER HASHING
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a hash of two digests and a domain identifier.
+    #[allow(dead_code)]
```

Comment:
This this is a `pub(crate)` trait, do we ever expect to use this internally? Otherwise we're just maintaining dead code

Reply:
I'm not sure we need this to marked as `dead_code`. I think `Rpo256::merge_in_domain()` should just be public and then all dead code warnings go away. I'll make a commit to fix this.

Reply:
Yep, missed that during the copy-paste. Thanks for fixing it!